### PR TITLE
Add a check that the declared search_fields exist

### PR DESCRIPTION
**Pull request by jschneier**
I got bitten by this when I was developing so thought I'd add a system check. Is this something you're interested in taking?

### CODE REVIEW

**Comment by jschneier** [comment]
We'll need to change the model definitions to put `index.Indexed` at the beginning as well. This is considered best practices for mixins I believe.

**Comment by jschneier** [comment]
I'm also unclear if I need to do more involved checks for `index.RelatedField`.

**Comment by jschneier** [comment]
I guess this PR also assumes that the mixin has been added to a model. Maybe should catch the `AttributeError` if a non-model is subclassed? Although I don't think the `super check` thing will work then.

File: wagtail/wagtailsearch/index.py [hunk 1]
```diff
@@ -77,6 +78,34 @@ def get_indexed_instance(self):
...
+            return hasattr(cls, name)
+
+    @classmethod
+    def check(cls, **kwargs):
+        errors = super(Indexed, cls).check(**kwargs)
+        errors.extend(cls._check_fields_exist(**kwargs))
+        return errors
+
+    @classmethod
+    def _check_fields_exist(cls, **kwargs):
```
Comment (kaedroho) [inline wagtail/wagtailsearch/index.py]: Could you rename this to ``_check_search_fields``?

File: wagtail/wagtailsearch/index.py [hunk 2]
```diff
@@ -77,6 +78,34 @@ def get_indexed_instance(self):
...
+        errors = []
+        for field in cls.get_search_fields():
+            message = '{name} in search_fields but not on {model}'
+            if not cls._has_field(field.field_name):
```
Comment (kaedroho) [inline wagtail/wagtailsearch/index.py]: This might be neater to implement on the ``BaseField`` class below as that class is responsible to doing the actual lookup. Could we move this to a new method on ``BaseField``? (eg ``BaseField.check(model)``)

File: wagtail/wagtailsearch/index.py [hunk 3]
```diff
@@ -77,6 +78,34 @@ def get_indexed_instance(self):
...
+    def check(cls, **kwargs):
+        errors = super(Indexed, cls).check(**kwargs)
+        errors.extend(cls._check_fields_exist(**kwargs))
+        return errors
+
+    @classmethod
+    def _check_fields_exist(cls, **kwargs):
+        errors = []
+        for field in cls.get_search_fields():
+            message = '{name} in search_fields but not on {model}'
```
Comment (kaedroho) [inline wagtail/wagtailsearch/index.py]: Maybe reword this message slightly? ``{model}.search_fields contains field '{name}' but it doesn't exist``

**Comment by kaedroho** [comment]
> I got bitten by this when I was developing so thought I'd add a system check. Is this something you're interested in taking?

Yep, definitely! Thanks for your work on this.

>  I'm also unclear if I need to do more involved checks for index.RelatedField.

It would be useful to have that there too. The ``RelatedFields`` class would need to check that each specified field exists on the related class as well.

The following should give an error:

```python

class Author(models.Model);
    name = models.CharField()

class Book(index.Indexed, models.Model):
    title = models.CharField()
    author = models.ForeignKey(Author)

    search_fields = [
        index.SearchField('title'),
        index.RelatedFields('author', [
            index.SearchField('name'),
            index.SearchField('bio'),   # Oops, bio doesn't exist on Author
        ]),
    ]
```

**Comment by jschneier** [comment]
Was there discussion around `index.Indexed` not being a `models.Model` subclass with `abstract = True`, because I am making a fair amount of assumptions about it being on a model in this PR I think. There are other things we can do like adding the check to a `checks.py` and registering it in that case.

**Comment by jschneier** [comment]
Given `.get_indexed_objects` I think the assumption is safe.

**Comment by jschneier** [comment]
@kaedroho can you have embedded `index.RelatedField` fields?

**Comment by kaedroho** [comment]
Yeah, that is allowed at the moment in Wagtail but I'm not sure if Elasticsearch supports it (their docs don't mention it and ES seems to just silently ignore anything that has a nested related field).

**Comment by gasman** [comment]
Does this correctly handle `SearchField`s on callables / other non-fields (http://docs.wagtail.io/en/v1.4.3/topics/search/indexing.html#indexing-callables-and-other-attributes)?

**Comment by jschneier** [comment]
Yes. It does not yet handle warning on Related Fields though, I need to find some time.

**Comment by kaedroho** [comment]
Thanks for your work on this so far @jschneier!

I've merged 720998c0cddacf9a4071b0c0e5952c7e17d8ce24 into master but I'd like to wait for the ``RelatedFields`` support before merging the rest in so bumping this PR to 1.6.

**Comment by gasman** [comment]
Have merged this (in 1dad100695060435b597545cd61c3ee420afee9f) as agreed with @kaedroho, as `RelatedFields` can be worked on separately. Thanks!
